blazar/utils/openstack/placement.py

```diff
@@ -427,3 +427,193 @@ def delete_reservation_inventory(self, host_name, reserv_uuid):
+
+        :param trait_name: The name of the trait
+        :raises: TraitCreationFailed error.
+        """
+        url = '/traits/%s' % trait_name
+        resp = self.put(url, {})
+        if resp:
+            LOG.info("Created trait %s", trait_name)
+            return
+        msg = ("Failed to create trait with placement API for "
+               "%(trait_name)s. Got %(status_code)d: %(err_text)s.")
+        args = {
+            'trait_name': trait_name,
+            'status_code': resp.status_code,
+            'err_text': resp.text,
+        }
+        LOG.error(msg, args)
+        raise exceptions.TraitCreationFailed(trait=trait_name)
+    
+    def delete_trait(self, trait_name):
+        """Calls the placement API to delete a trait.
+
+        :param trait_name: The name of the trait
+        :raises: TraitDeletionFailed error.
+        """
+        url = '/traits/%s' % trait_name
+        resp = self.delete(url)
+        if resp:
+            LOG.info("Deleted trait %s", trait_name)
+            return
+        msg = ("Failed to delete trait with placement API for "
+               "%(trait_name)s. Got %(status_code)d: %(err_text)s.")
+        args = {
+            'trait_name': trait_name,
+            'status_code': resp.status_code,
+            'err_text': resp.text,
+        }
+        LOG.error(msg, args)
+        raise exceptions.TraitDeletionFailed(trait=trait_name)
+        
+    def create_reservation_trait(self, reserv_uuid, project_id):
+        """Create the reservation trait.
+        
+        :param reservation_uuid: The reservation uuid
+        :raises: TraitCreationFailed error.
+        """
+        self.create_trait(self._get_custom_reservation_trait_name(reserv_uuid, project_id))
+    
+    def delete_reservation_trait(self, reserv_uuid, project_id):
+        """Delete the reservation trait.
+        
+        :param reservation_uuid: The reservation uuid
+        :raises: TraitDeletionFailed error.
+        """
+        self.delete_trait(self._get_custom_reservation_trait_name(reserv_uuid, project_id))
+    
+    def associate_traits_with_resource_provider(self, rp_uuid, traits):
+        """Associate traits with the resource provider.
+        
+        :param rp_uuid: The uuid of the resource provider
+        :param traits: The list of traits
+        :raises: TraitAssociationFailed error.
+        """
+        url = '/resource_providers/%s/traits' % rp_uuid
+        resp = self.get(url)
+        current_traits = []
+        resource_provider_generation = 0
+        if resp:
+            json_resp = resp.json()
+            if json_resp['traits']:
+                current_traits = json_resp['traits']
+            if json_resp['resource_provider_generation']:
+                resource_provider_generation = json_resp['resource_provider_generation']
+                
+        updated_traits = current_traits + [trait for trait in traits if trait not in current_traits]
```

Comment:
You can use sets to make this a bit cleaner: `list(set(current_traits) + set(traits))`